hw/syn/tools/dc/run-syn.tcl

```diff
@@ -0,0 +1,120 @@
+
+#######################
+##  DESIGN SOURCES  ###
+#######################
+
+# just compile the "core" toplevel at the moment
+# might want to switch to top_earlgrey_asic later on (with pads)
+set TOP_ENTITY [getenv "TOP_ENTITY"]
+
+# read source list generated by fusesoc
+set fp [open [ls "*.scr"] r]
+set file_data [read $fp]
+close $fp
+set data [split $file_data "\n"]
+
+# this TECH_LIB_SEL selects the appropriate technology by defining
+# PRIM_DEFAULT_IMPL=prim_pkg::Impl<tech identifier>
+# TECH_LIB_SEL is set inside the library setup script
+set DEFINE "PRIM_DEFAULT_IMPL=${PRIM_DEFAULT_IMPL} "
+# go through fusesoc file list and separate +defines from actual source files
+set SRC {}
+foreach item $data {
+  if {[string range $item 0 7] == "+define+"}  {
+    set DEFINE "${DEFINE}[string range $item 8 end] "
+  } elseif {[string range $item 0 7] == "+incdir+"}  {
+    lappend search_path "[string range $item 8 end]"
+  } else {
+    lappend SRC $item
+  }
+}
+
+# additional parameters
+set PARAMS ""
+
+###########################
+##   ELABORATE DESIGN    ##
+###########################
+
+# delete previous designs.
+remove_design -designs
+sh rm -rf $WORKLIB/*
+
+analyze -define ${DEFINE} -format sv ${SRC}
+elaborate  ${TOP_ENTITY} -parameters ${PARAMS}
+link > "${REPDIR}/${TOP_ENTITY}_link.rpt"
+
+write_file -format ddc -hierarchy -output "${DDCDIR}/${TOP_ENTITY}_elab.ddc"
+
+#############################
+##   CLOCK GATING SETUP    ##
+#############################
+
+# be more specific if defaults do not suffice
+# set_clock_gating_style -num_stages 1                   \
+#                        -positive_edge_logic integrated \
+#                        -control_point before           \
+#                        -control_signal scan_enable
+
+###########################
+##   APPLY CONSTRAINTS   ##
+###########################
+
+source constraints.sdc
+
+# If hold time should be fixed
+# set_fix_hold ${CLK_PIN}
+
+######################
+##    MAP DESIGN    ##
+######################
+
+# TODO: we may have to disable a couple of optimizations in order
+# to prevent the tool from optimizing away dummy logic or logic from blocks
+# that are only half-finished
+
+# preserve hierarchy for reports
+compile_ultra -gate_clock -scan -no_autoungroup
+
+#################
+##   REPORTS   ##
+#################
+
+report_timing      -nosplit -nworst 100 > "${REPDIR}/${TOP_ENTITY}_timing.rpt"
+report_area  -hier -nosplit             > "${REPDIR}/${TOP_ENTITY}_area.rpt"
+report_power -hier -nosplit             > "${REPDIR}/${TOP_ENTITY}_power.rpt"
+
+#################
+##   NETLIST   ##
+#################
+
+change_names -rules verilog -hierarchy
+# define_name_rules fixbackslashes -allowed "A-Za-z0-9_" -first_restricted "\\" -remove_chars
+# change_names -rule fixbackslashes -h
+write_file -format ddc     -hierarchy -output "${DDCDIR}/${TOP_ENTITY}_mapped.ddc"
+write_file -format verilog -hierarchy -output "${VLOGDIR}/${TOP_ENTITY}_mapped.v"
+
```

Comment:
if we need it for any reason, this would be the time to flatten and then write_file ddc/verilog of the flattened output as well.  Sometimes you get better QOR doing a flatten, compile_ultra -incr too (even though it is supposed to do something like this under the hood)

Reply:
Yeah so I only keep the hierarchy for reporting reasons, since this flow is really just to monitor the area distribution among sub modules and make sure that timing looks fine.

For an actual "production" synthesis I would of course also consider flattening individual comportable IPs, or more.

But I can provision the incremental step here.